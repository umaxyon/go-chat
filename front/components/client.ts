import { split, HttpLink, ApolloClient, InMemoryCache, gql } from '@apollo/client'
import { GraphQLWsLink } from '@apollo/client/link/subscriptions'
import { createClient } from 'graphql-ws'
import { getMainDefinition } from '@apollo/client/utilities'

const httpLink = new HttpLink({
    uri: 'http://localhost:8080/query'
})

const wsLink = typeof window !== "undefined" ? new GraphQLWsLink(createClient({
    url: 'ws://localhost:8080/query',
})) : httpLink;

const splitLink = split(({ query }) => {
    const definition = getMainDefinition(query)
    return (definition.kind === 'OperationDefinition' && definition.operation === 'subscription')
}, wsLink, httpLink)

export const client = new ApolloClient({ link: splitLink, cache: new InMemoryCache() })

export const ADD_COMMENT = gql`
mutation($user: String!, $text: String!) {
  postMessage(user: $user, text: $text) {
    id
    user
    text
    createdAt
  }
}
`