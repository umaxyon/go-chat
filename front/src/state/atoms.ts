import { atom } from "recoil"

export type Member = {
    user: string
}

export type FeedRow = {
    id: string,
    MessageType: "comment" | "addMember" | "leaveMember" | "system",
    user: string,
    text: string,
    createdAt: any
}

export type LoginData = {
    user: string,
    token: string,
    disconnect?: () => void
}

export const membersState = atom<Member[]>({
    key: "members",
    default: [],
})

export const commentsState = atom<FeedRow[]>({
    key: "comments",
    default: [],
})

export const loginState = atom<LoginData>({
    key: "loginUser",
    default: undefined,
})