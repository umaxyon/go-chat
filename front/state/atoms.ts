import { atom } from "recoil"

export type Member = {
    user: string
}

export type Comment = {
    id: string,
    user: string,
    text: string,
    createdAt: any
}

export type LoginData = {
    user: string,
    disconnect?: () => void
}

export const membersState = atom<Member[]>({
    key: "members",
    default: [],
})

export const commentsState = atom<Comment[]>({
    key: "comments",
    default: [],
})

export const loginState = atom<LoginData>({
    key: "loginUser",
    default: undefined,
})