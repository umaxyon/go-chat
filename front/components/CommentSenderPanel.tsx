import { useMutation } from "@apollo/client"
import { ChangeEventHandler, useState } from "react"
import { ADD_COMMENT } from "./client"

type CommentSenderPanelProps = {
    user: string
}

const CommentSenderPanel: React.FC<CommentSenderPanelProps> = ({ user }) => {
    const [ addComment, { error } ] = useMutation(ADD_COMMENT)
    const [ comment, setComment ] = useState<string>("")

    const onChangeComment: ChangeEventHandler<HTMLInputElement> = (e) => {
        setComment(e.target.value)
    }

    const onClickSubmit = async () => {
        await addComment({ variables: { user, text: comment }})
    }

    return (
        <>
        { error ? <span>{error.message}</span>: null }
        <div className="flex justify-center gap-2">
            <input type="text" placeholder='入力してください' onChange={onChangeComment}
            className="
                block w-full text-base font-normal text-gray-700 bg-white bg-clip-padding
                border border-solid border-gray-300 rounded py-1.5 px-3 transition ease-in-out m-0
                focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none" />
            <button type="button" onClick={onClickSubmit}
            className="
                inline-block px-6 py-2.5 bg-blue-400 text-white font-medium text-xs leading-tight uppercase whitespace-nowrap
                rounded shadow-md hover:bg-blue-500 hover:shadow-lg focus:bg-blue-500 focus:shadow-lg focus:outline-none 
                focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out">送信</button>
      </div>
      </>
    )
}
export default CommentSenderPanel