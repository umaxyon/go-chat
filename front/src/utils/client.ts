import fetch from 'cross-fetch';
import { split, HttpLink, ApolloClient, InMemoryCache, gql } from '@apollo/client'
import { GraphQLWsLink } from '@apollo/client/link/subscriptions'
import { createClient } from 'graphql-ws'
import { getMainDefinition } from '@apollo/client/utilities'

const httpLink = new HttpLink({
    uri: 'http://localhost:8080/query', fetch
})

const wsLink = typeof window !== "undefined" ? new GraphQLWsLink(createClient({
    url: 'ws://localhost:8080/query',
})) : httpLink;

const splitLink = split(({ query }) => {
    const definition = getMainDefinition(query)
    return (definition.kind === 'OperationDefinition' && definition.operation === 'subscription')
}, wsLink, httpLink)

export const client = new ApolloClient({ link: splitLink, cache: new InMemoryCache() })

export const ADD_COMMENT = gql`
mutation($user: String!, $text: String!, $token: String!) {
  postMessage(user: $user, text: $text, token: $token) {
    id
    MessageType
    user
    text
    createdAt
  }
}
`

export const INIT_LOAD_QUERY = gql`
query GetResponse {
  messages {
    id
    MessageType
    user
    text
    createdAt
  }
  members {
    user
  }
}
`
export const SUBSCRIPTION = gql`
subscription subscribeMessage($user: String!) {
  subscribe(user: $user) {
    message {
        id
        MessageType
        user
        text
        createdAt
    }
    user {
        user
    }
    leave {
      user
    }
  }
}
`