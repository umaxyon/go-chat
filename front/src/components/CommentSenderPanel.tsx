import { useMutation } from "@apollo/client"
import { ChangeEventHandler, KeyboardEventHandler, useCallback, useEffect, useRef, useState } from "react"
import { useRecoilValue } from "recoil"
import { useLogout } from "../hooks/logout"
import { loginState } from "../state/atoms"
import { ADD_COMMENT } from "../utils/client"

type CommentSenderPanelProps = {
}

const CommentSenderPanel: React.FC<CommentSenderPanelProps> = () => {
    const txtInput = useRef<HTMLInputElement>(null)
    const [ addComment, { error } ] = useMutation(ADD_COMMENT)
    const [ comment, setComment ] = useState<string>("")
    const login = useRecoilValue(loginState)
    const logout = useLogout()

    useEffect(() => txtInput.current?.focus())

    const onChangeComment: ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
        setComment(e.target.value)
    }, [])

    const onClickSubmit = useCallback(async () => {
        if (comment) {
            try {
                await addComment({ variables: { ...login, text: comment }})
                setComment("")
            } catch {
                logout()
            }
        }
    }, [comment, login, addComment, logout])

    const onKeyUp: KeyboardEventHandler<HTMLInputElement> = useCallback(async (e) => {
        if (e.key === 'Enter') {
            await onClickSubmit()
        }
    }, [onClickSubmit])

    return (
        <>
        { error ? <span>{error.message}</span>: null }
        <div className="flex justify-center gap-2">
            <input type="text" placeholder='入力してください' onChange={onChangeComment} onKeyUp={onKeyUp} value={comment} ref={txtInput}
            className="
                block w-full text-base font-normal text-gray-700 bg-white bg-clip-padding
                border border-solid border-gray-300 rounded py-1.5 px-3 transition ease-in-out m-0
                focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none" />
            <button type="button" onClick={onClickSubmit}
            className="
                inline-block px-6 py-2.5 bg-blue-400 text-white font-medium text-xs leading-tight uppercase whitespace-nowrap
                rounded shadow-md hover:bg-blue-500 hover:shadow-lg focus:bg-blue-500 focus:shadow-lg focus:outline-none 
                focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out">送信</button>
        </div>
        </>
    )
}
export default CommentSenderPanel