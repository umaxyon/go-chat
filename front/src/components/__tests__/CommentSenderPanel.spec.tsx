import CommentSenderPanel from "@/components/CommentSenderPanel"
import { MockedProvider } from '@apollo/client/testing';
import "@testing-library/jest-dom/extend-expect";
import { cleanup, screen, fireEvent, render } from '@testing-library/react';
import React from "react";
import renderer from "react-test-renderer"
import { useRecoilState, useRecoilValue } from "recoil";
import { LoginData } from "../../state/atoms";


afterEach(() => cleanup())
const setRecoilValueMock = (val: LoginData) => {
    (useRecoilValue as jest.Mock).mockImplementation(() => val)
}
const setRecoilStateMock = (state: any[]) => {
    (useRecoilState as jest.Mock).mockImplementation(() => state)
}

jest.mock("recoil");

const TEST_TEXT  = "abcde"
const TEST_USER  = "dog"
const TEST_TOKEN = "dummy"

const sut = (
    <MockedProvider addTypename={false}>
        <CommentSenderPanel />
    </MockedProvider>
)

it('snapshot test', () => {
    setRecoilStateMock([ "", jest.mock ])
    const component = renderer.create(sut);
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

describe('CommentSenderPanelテスト', () => {
    // let useStateSpy: jest.SpyInstance
    let setComment : jest.Mock

    // const setCommentSpySetup = (initial: string) => {
    //     useStateSpy.mockImplementation((): any => [initial, setComment] as any)
    // }

    beforeEach(() => {
        setComment = jest.fn()
        // useStateSpy = jest.spyOn(React, "useState")
        // setCommentSpySetup("")
        setRecoilValueMock({ user: TEST_USER, token: TEST_TOKEN })
        setRecoilStateMock([ "", setComment ])
    })

    it('テキストボックスに入力してボタン押すとsetCommentされてボックスがクリアされる', async () => {
        render(sut)
        const textbox = screen.getByRole('textbox')
        expect(textbox).toHaveValue("")
        fireEvent.change(textbox, { target: { value: TEST_TEXT}})
        fireEvent.click(screen.getByRole('button'))
        expect(setComment).toHaveBeenLastCalledWith(TEST_TEXT)
        expect(screen.getByRole('textbox')).toHaveValue("")
    })

    it('テキストボックスに入力してEnterキーを押すとsetCommentされてボックスがクリアされる', () => {
        render(sut)
        const textbox = screen.getByRole('textbox')
        expect(textbox).toHaveValue("")
        fireEvent.change(textbox, { target: { value: TEST_TEXT}})
        fireEvent.keyUp(textbox, { target: { key: "Enter" }})
        expect(setComment).toHaveBeenLastCalledWith(TEST_TEXT)
        expect(screen.getByRole('textbox')).toHaveValue("")
    })

    it('テキストボックスが空でボタンを押すとsetCommentが呼ばれない', () => {
        render(sut)
        fireEvent.click(screen.getByRole("button"))
        expect(setComment).not.toHaveBeenCalled()
    })
    
    it('テキストボックスが空でEnterを押すとsetCommentが呼ばれない', () => {
        render(sut)
        fireEvent.keyUp(screen.getByRole('textbox'), { target: { key: "Enter" }})
        expect(setComment).not.toHaveBeenCalled()
    })
})
