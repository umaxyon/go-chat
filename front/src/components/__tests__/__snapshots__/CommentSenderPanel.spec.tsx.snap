// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="flex justify-center gap-2"
>
  <textarea
    className=" block w-full text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded py-1.5 px-3 transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="入力してください (ctrl + enter 送信)"
    value=""
  />
  <button
    className=" inline-block px-6 py-2.5 bg-blue-400 text-white font-medium text-xs leading-tight uppercase whitespace-nowrap rounded shadow-md hover:bg-blue-500 hover:shadow-lg focus:bg-blue-500 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out"
    onClick={[Function]}
    type="button"
  >
    送信
  </button>
</div>
`;
